draft-ietf-mls-protocol.md

```diff
@@ -2882,83 +2897,59 @@ struct {
 MUST verify that the `version` field of the KeyPackage has the same value as the
 `version` field of the MLSMessage.  The `version` field in the KeyPackage
 provides an explicit signal of the intended version to the other members of
 group when they receive the KeyPackage in an Add proposal.
 
-The `capabilities` field in the `leaf_node` allows MLS session
+The `capabilties` field in the `client_info` allows MLS session
 establishment to be safe from downgrade attacks on the parameters described (as
 discussed in {{group-creation}}), while still only advertising one version /
 ciphersuite per KeyPackage.
 
-The `leaf_node_source` field of the LeafNode in a KeyPackage MUST be set to
+The `client_info.client_info_source` field in a KeyPackage MUST be set to
 `key_package`.
 
-## KeyPackage Validation
-
-The validity of a KeyPackage needs to be verified at a few stages:
+Extensions that appear in the KeyPackage `client_info.extensions` field
+MUST be included in the `client_info.capabilities` field.
 
-* When a KeyPackage is downloaded by a group member, before it is used
-  to add the client to the group
-* When a KeyPackage is received by a group member in an Add message
+## KeyPackage Validation
 
+The validity of a KeyPackage needs to be verified when a KeyPackage is
+downloaded by a group member, before it is used to add the client to the group.
 The client verifies the validity of a KeyPackage using the following steps:
 
-* Verify the LeafNode in the KeyPackage according to the process defined in
+* Verify `client_info` in the KeyPackage according to the process defined in
   {{leaf-node-validation}}.
 
 * Verify that the signature on the KeyPackage is valid using the public key
-  in the LeafNode's credential
+  in the `client_info.credential`
 
-* Verify that the ciphersuite and protocol version of the LeafNode match
+* Verify that the ciphersuite and protocol version of the KeyPackage match
   those in use in the group.
 
-## Client Capabilities
-
-The `capabilities` extension indicates what protocol versions, ciphersuites,
-protocol extensions, and non-default proposal types are supported by a client.
-Proposal types defined in this document are considered "default" and thus need
-not be listed.
-
-~~~~~
-struct {
-    ProtocolVersion versions<V>;
-    CipherSuite ciphersuites<V>;
-    ExtensionType extensions<V>;
-    ProposalType proposals<V>;
-} Capabilities;
-~~~~~
-
-This extension MUST be always present in a KeyPackage.  Extensions that appear
-in the `extensions` field of a KeyPackage MUST be included in the `extensions`
-field of the `capabilities` extension.
-
 ## Lifetime
 
-The `lifetime` extension represents the times between which clients will
-consider a KeyPackage valid.  This time is represented as an absolute time,
-measured in seconds since the Unix epoch (1970-01-01T00:00:00Z). A client MUST
+The `client_info.lifetime` field represents the times between which clients will
+consider a KeyPackage valid.  A client MUST
```

Comment:
This trails off.

Reply:
In any case, this section should be deleted, because Lifetime isn't an extension any more.